File: obv_messenger/app/src/main/java/io/olvid/messenger/fragments/dialog/LockScreenPINCreationDialogFragment.java
/*
 *  Olvid for Android
 *  Copyright © 2019-2022 Olvid SAS
 *
 *  This file is part of Olvid for Android.
 *
 *  Olvid is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License, version 3,
 *  as published by the Free Software Foundation.
 *
 *  Olvid is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with Olvid.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.olvid.messenger.fragments.dialog;

import android.app.Dialog;
import android.os.Bundle;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;

import io.olvid.messenger.App;
import io.olvid.messenger.R;
import io.olvid.messenger.customClasses.TextChangeListener;
import io.olvid.messenger.settings.SettingsActivity;


public class LockScreenPINCreationDialogFragment extends DialogFragment implements View.OnClickListener {
    private Runnable onPINSetCallback = null;
    private boolean usePassword = false;

    private Button createPINButton;
    private EditText pinFirstInput;
    private EditText pinSecondInput;
    private Button switchPasswordPIN;
    private TextView textPINErrorMessage;

    public static LockScreenPINCreationDialogFragment newInstance() {
        return new LockScreenPINCreationDialogFragment();
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        Dialog dialog = super.onCreateDialog(savedInstanceState);

        Window window = dialog.getWindow();
        if (window != null) {
            window.requestFeature(Window.FEATURE_NO_TITLE);
            if (SettingsActivity.preventScreenCapture()) {
                window.setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE);
            }
        }
        return dialog;
    }

    @Override
    public void onStart() {
        super.onStart();
        Dialog dialog = getDialog();
        if (dialog != null) {
            Window window = dialog.getWindow();
            if (window != null) {
                window.setLayout(WindowManager.LayoutParams.MATCH_PARENT, WindowManager.LayoutParams.WRAP_CONTENT);
                window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE | WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
                pinFirstInput.requestFocus();
            }
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View dialogView = inflater.inflate(R.layout.dialog_fragment_lock_screen_pin_creation, container, false);

        createPINButton = dialogView.findViewById(R.id.button_create_pin);
        createPINButton.setOnClickListener(this);

        Button cancelButton = dialogView.findViewById(R.id.button_cancel);
        cancelButton.setOnClickListener(this);

        switchPasswordPIN = dialogView.findViewById(R.id.button_switch_password_pin);
        switchPasswordPIN.setOnClickListener(this);

        textPINErrorMessage = dialogView.findViewById(R.id.text_pin_error_message);

        pinFirstInput = dialogView.findViewById(R.id.pin_first_input);
        pinSecondInput = dialogView.findViewById(R.id.pin_second_input);
        TextWatcher textWatcher = new TextChangeListener() {
            @Override
            public void afterTextChanged(Editable s) {
                validatePINs();
            }
        };
        pinFirstInput.addTextChangedListener(textWatcher);
        pinSecondInput.addTextChangedListener(textWatcher);

        usePassword = SettingsActivity.isPINAPassword();
        configureInputsForPINOrPassword();
        return dialogView;
    }

    private void validatePINs() {
        String firstPIN = pinFirstInput.getText().toString();
        if (firstPIN.length() < 4) {
            if (firstPIN.length() != 0) {
                if (usePassword) {
                    textPINErrorMessage.setText(R.string.error_text_password_too_short);
                } else {
                    textPINErrorMessage.setText(R.string.error_text_pin_too_short);
                }
            } else {
                textPINErrorMessage.setText(null);
            }
            pinSecondInput.setEnabled(false);
            createPINButton.setEnabled(false);
            return;
        } else {
            textPINErrorMessage.setText(null);
            pinSecondInput.setEnabled(true);
        }

        String secondPIN = pinSecondInput.getText().toString();
        if (!firstPIN.equals(secondPIN)) {
            if (secondPIN.length() > 0) {
                if (usePassword) {
                    textPINErrorMessage.setText(R.string.error_text_password_mismatch);
                } else {
                    textPINErrorMessage.setText(R.string.error_text_pin_mismatch);
                }
            } else {
                textPINErrorMessage.setText(null);
            }
            createPINButton.setEnabled(false);
        } else {
            textPINErrorMessage.setText(null);
            createPINButton.setEnabled(true);
        }
    }

    public void setOnPINSetCallBack(Runnable callBack) {
        this.onPINSetCallback = callBack;
    }

    private void configureInputsForPINOrPassword() {
        if (usePassword) {
            pinFirstInput.setText(null);
            pinFirstInput.setHint(R.string.hint_enter_password);
            pinFirstInput.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
            pinSecondInput.setText(null);
            pinSecondInput.setHint(R.string.hint_confirm_password);
            pinSecondInput.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
            switchPasswordPIN.setText(R.string.button_label_use_pin);
        } else {
            pinFirstInput.setText(null);
            pinFirstInput.setHint(R.string.hint_enter_pin);
            pinFirstInput.setInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_VARIATION_PASSWORD);
            pinSecondInput.setText(null);
            pinSecondInput.setHint(R.string.hint_confirm_pin);
            pinSecondInput.setInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_VARIATION_PASSWORD);
            switchPasswordPIN.setText(R.string.button_label_use_password);
        }
    }

    @Override
    public void onClick(View view) {
        int id = view.getId();
        if (id == R.id.button_create_pin) {
            String firstPIN = pinFirstInput.getText().toString();
            String secondPIN = pinSecondInput.getText().toString();
            if (firstPIN.equals(secondPIN)) {
                SettingsActivity.savePIN(firstPIN, usePassword);
                dismiss();
                if (usePassword) {
                    App.toast(R.string.toast_message_new_password_set, Toast.LENGTH_SHORT);
                } else {
                    App.toast(R.string.toast_message_new_pin_set, Toast.LENGTH_SHORT);
                }
                if (onPINSetCallback != null) {
                    onPINSetCallback.run();
                }
            }
        } else if (id == R.id.button_cancel) {
            dismiss();
        } else if (id == R.id.button_switch_password_pin) {
            usePassword = !usePassword;
            configureInputsForPINOrPassword();
        }
    }
}

File: obv_messenger/app/src/main/java/io/olvid/messenger/lock_screen/LockScreenPINCreationDialogFragment.kt
/*
 *  Olvid for Android
 *  Copyright © 2019-2026 Olvid SAS
 *
 *  This file is part of Olvid for Android.
 *
 *  Olvid is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License, version 3,
 *  as published by the Free Software Foundation.
 *
 *  Olvid is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with Olvid.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.olvid.messenger.lock_screen

import android.app.Dialog
import android.content.res.Configuration
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowManager
import android.widget.Toast
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.fragment.app.DialogFragment
import io.olvid.messenger.App
import io.olvid.messenger.R
import io.olvid.messenger.customClasses.formatMarkdownToAnnotatedString
import io.olvid.messenger.designsystem.components.BaseDialogContent
import io.olvid.messenger.designsystem.components.OlvidActionButton
import io.olvid.messenger.designsystem.components.OlvidPasswordInput
import io.olvid.messenger.designsystem.components.OlvidTextButton
import io.olvid.messenger.designsystem.cutoutHorizontalPadding
import io.olvid.messenger.designsystem.systemBarsHorizontalPadding
import io.olvid.messenger.designsystem.theme.OlvidTypography
import io.olvid.messenger.settings.SettingsActivity

class LockScreenPINCreationDialogFragment : DialogFragment() {
    private var onPINSetCallback: Runnable? = null

    companion object {
        @JvmStatic
        fun newInstance() = LockScreenPINCreationDialogFragment()
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState)
        dialog.window?.let { window ->
            window.requestFeature(Window.FEATURE_NO_TITLE)
            if (SettingsActivity.preventScreenCapture(context)) {
                window.setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE)
            }
        }
        return dialog
    }

    override fun onStart() {
        super.onStart()
        dialog?.window?.apply {
            setLayout(WindowManager.LayoutParams.MATCH_PARENT, WindowManager.LayoutParams.WRAP_CONTENT)
            setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE or WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)
            setBackgroundDrawableResource(android.R.color.transparent)
        }
    }

    fun setOnPINSetCallBack(callBack: Runnable) {
        onPINSetCallback = callBack
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return ComposeView(inflater.context).apply {
            setContent {
                Box(
                    contentAlignment = Alignment.Center
                ) {
                    PINCreationContent(
                        isPinAPassword = SettingsActivity.isPINAPassword,
                        onCancel = { dismiss() },
                        onPINSet = { pin, isPassword ->
                            SettingsActivity.savePIN(pin = pin, pinIsAPassword = isPassword)
                            dismiss()
                            App.toast(
                                if (isPassword) R.string.toast_message_new_password_set else R.string.toast_message_new_pin_set,
                                Toast.LENGTH_SHORT
                            )
                            onPINSetCallback?.run()
                        }
                    )
                }
            }
        }
    }
}

@Composable
private fun PINCreationContent(
    isPinAPassword: Boolean,
    onCancel: () -> Unit,
    onPINSet: (String, Boolean) -> Unit,
) {
    var usePassword by remember { mutableStateOf(isPinAPassword) }
    val firstPin = remember { mutableStateOf("") }
    val secondPin = remember { mutableStateOf("") }

    val firstFocusRequester = remember { FocusRequester() }
    val secondFocusRequester = remember { FocusRequester() }

    val firstTooShort = firstPin.value.isNotEmpty() && firstPin.value.length < 4
    val secondEnabled = firstPin.value.length >= 4
    val pinsMismatch = secondEnabled && secondPin.value.isNotEmpty() && firstPin.value != secondPin.value
    val createEnabled = secondEnabled && firstPin.value == secondPin.value

    val errorMessage = when {
        firstTooShort -> stringResource(if (usePassword) R.string.error_text_password_too_short else R.string.error_text_pin_too_short)
        pinsMismatch -> stringResource(if (usePassword) R.string.error_text_password_mismatch else R.string.error_text_pin_mismatch)
        else -> ""
    }

    LaunchedEffect(Unit) {
        firstFocusRequester.requestFocus()
    }

    BaseDialogContent(
        modifier = Modifier
            .cutoutHorizontalPadding()
            .systemBarsHorizontalPadding()
            .padding(horizontal = 16.dp, vertical = 4.dp)
            .widthIn(max = 320.dp),
        title = stringResource(R.string.dialog_title_pin_creation),
        content = {
            Text(
                modifier = Modifier.padding(bottom = 8.dp),
                text = stringResource(R.string.text_pin_creation_explanation).formatMarkdownToAnnotatedString(),
                style = OlvidTypography.body2,
                color = colorResource(R.color.greyTint),
            )

            OlvidPasswordInput(
                modifier = Modifier
                    .fillMaxWidth()
                    .focusRequester(firstFocusRequester),
                label = stringResource(if (usePassword) R.string.hint_enter_password else R.string.hint_enter_pin),
                password = firstPin,
                keyboardOptions = KeyboardOptions(
                    keyboardType = if (usePassword) KeyboardType.Password else KeyboardType.NumberPassword,
                    imeAction = ImeAction.Next,
                ),
                keyboardActions = KeyboardActions(
                    onNext = { if (secondEnabled) secondFocusRequester.requestFocus() }
                ),
                isError = firstTooShort
            )

            Spacer(Modifier.height(8.dp))

            OlvidPasswordInput(
                modifier = Modifier
                    .fillMaxWidth()
                    .focusRequester(secondFocusRequester),
                label = stringResource(if (usePassword) R.string.hint_confirm_password else R.string.hint_confirm_pin),
                password = secondPin,
                enabled = secondEnabled,
                keyboardOptions = KeyboardOptions(
                    keyboardType = if (usePassword) KeyboardType.Password else KeyboardType.NumberPassword,
                    imeAction = ImeAction.Done,
                ),
                keyboardActions = KeyboardActions(
                    onDone = { if (createEnabled) onPINSet(firstPin.value, usePassword) }
                ),
                isError = pinsMismatch,
            )

            // always rendered to avoid layout shifts
            Text(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 4.dp),
                text = errorMessage,
                style = OlvidTypography.body2,
                color = colorResource(R.color.red),
            )

            OlvidTextButton(
                modifier = Modifier.align(Alignment.End).height(40.dp),
                text = stringResource(if (usePassword) R.string.button_label_use_pin else R.string.button_label_use_password),
                onClick = {
                    usePassword = !usePassword
                    firstPin.value = ""
                    secondPin.value = ""
                }
            )
        },
        actions = {
            Spacer(Modifier.weight(1f, true))
            OlvidTextButton(
                text = stringResource(R.string.button_label_cancel),
                onClick = onCancel,
            )
            Spacer(Modifier.width(8.dp))
            OlvidActionButton(
                text = stringResource(R.string.button_label_create_pin),
                enabled = createEnabled,
                onClick = {
                    onPINSet(firstPin.value, usePassword)
                },
            )
        }
    )
}


@Preview
@Preview(uiMode = Configuration.UI_MODE_NIGHT_YES or Configuration.UI_MODE_TYPE_NORMAL,
    locale = "fr"
)
@Composable
private fun PINCreationContentPreview() {
    PINCreationContent(
        isPinAPassword = false,
        onCancel = {},
        onPINSet = { _, _ ->

        },
    )
}